src/main/scala/org/apache/spark/mllib/linalg/distributed/BetterBlockMatrix.scala

```diff
@@ -0,0 +1,183 @@
+    val rPartitioner = r.blocks.partitioner.get.asInstanceOf[GridPartitioner]
+    val nRows: Long = l.numRows()
+    val nCols: Long = r.numCols()
+    val rowsPerBlock: Int = l.rowsPerBlock
+    val colsPerBlock: Int = r.colsPerBlock
+    val rowBlocks = l.numRowBlocks
+    val colBlocks = r.numColBlocks
+    val nProducts = l.numColBlocks
+    val rowsRemainder: Int = (nRows % rowsPerBlock).toInt
+    val colsRemainder: Int = (nCols % colsPerBlock).toInt
+
+    private def toDenseMatrix(x: Matrix): DenseMatrix = x match {
+      case x: DenseMatrix => x
+      case x: SparseMatrix => x.toDense
+    }
+
+    private def fmaDenseMatricesWithExtension(x: DenseMatrix, y: DenseMatrix, result: DenseMatrix) {
+      val x2 = if (result.numRows != x.numRows)
+        toDenseMatrix(Matrices.vertcat(Array(x, Matrices.zeros(result.numRows - x.numRows, x.numCols))))
+      else
+        x
+
+      val y2 = if (result.numCols != y.numCols)
+        toDenseMatrix(Matrices.horzcat(Array(y, Matrices.zeros(y.numRows, result.numCols - y.numCols))))
+      else
+        y
+
+      BLAS.gemm(1.0, x2, y2, 1.0, result)
+    }
+
+    private def block(bm: BlockMatrix, p: GridPartitioner, context: TaskContext, i: Int, j: Int): Matrix =
+      bm.blocks.compute(IntPartition(p.getPartition((i, j))), context).toArray match {
+        case Array((_, m)) => m
+        case x => throw new RuntimeException(s"Expected array of length one, but got blocks: ${x.map(_._1).toSeq}, from $bm with partitioner $p.")
+      }
+    private def leftBlock(i: Int, j: Int, context: TaskContext): Matrix =
+      block(l, lPartitioner, context, i, j)
+    private def rightBlock(i: Int, j: Int, context: TaskContext): Matrix =
+      block(r, rPartitioner, context, i, j)
+
+    def compute(split: Partition, context: TaskContext): Iterator[((Int, Int), Matrix)] = {
+      val row = split.index % rowBlocks
+      val col = split.index / rowBlocks
+      val rowsInThisBlock: Int = (if (row + 1 == rowBlocks && rowsRemainder != 0) rowsRemainder else rowsPerBlock)
+      val colsInThisBlock: Int = (if (col + 1 == colBlocks && colsRemainder != 0) colsRemainder else colsPerBlock)
+
+      val finalResult = DenseMatrix.zeros(rowsInThisBlock, colsInThisBlock)
+      var i = 0
+      while (i < nProducts) {
+        val leftMat = leftBlock(row, i, context)
+        val rightMat = rightBlock(i, col, context)
+        (leftMat, rightMat) match {
+          case (x: DenseMatrix, y: DenseMatrix) =>
+            fmaDenseMatricesWithExtension(x, y, finalResult)
+          case (x: DenseMatrix, y: SparseMatrix) =>
+            fmaDenseMatricesWithExtension(x, toDenseMatrix(y), finalResult)
+          case (x: SparseMatrix, y: DenseMatrix) =>
+            fmaDenseMatricesWithExtension(toDenseMatrix(x), y, finalResult)
+          case _ =>
+             throw new SparkException(s"I only support Dense * Dense, recieved: ${leftMat.getClass}, ${rightMat.getClass}.")
```

Comment:
Also, it's at the very least an untrue error message, since we clearly support Dense * Sparse and Sparse * Dense. Replace "I" in "I only support" with something more descriptive. Also, there's a typo in "received"